### PR TITLE
snapping for selections and vertical space

### PR DESCRIPTION
**Pull request by rolandlo**
This first commit implements snapping for moving selections, addressing issue #807 (and the duplicate #1972). The selection is snapped according to the the nearest corner of the reduced bounding box (discarding padding and stroke width). This still needs some fixes (in particular for scaling/rotating objects and for selections which do not merely consist of strokes) and improvements.

I do plan to add snapping for vertical space (mentioned in issue #807).

### CODE REVIEW

**Comment by rolandlo** [comment]
Currently snapping for moving selections looks like in the demo video below:

snappingDemo

**Comment by rolandlo** [comment]
Now snapping works for selections consisting only of strokes, even when they have been scaled and rotated. Moreover you can now make selections as small as minSize (5 pixels in view coordinates) and do not lose the aspect ratio while scaling them up again.

A few things left:
- If the selection contains other elements than strokes, the snapping does not work properly yet.
- While the rotation angle is non-zero, resizing the selection results in a additional translation movement. This is due to the fact that resizing results in a change of the rotation center (two rotations with different centers and opposite angle combined yield a translation). This is not new and is not so problematic, because the selection can quickly be moved and snapped again.
- The cursor for resizing should be rotated along with the selection (issue #2021)
- It might be nice to have snapping for resizing selections as well. However, I am not sure, what the best behavior is, in particular for rotated selections. 
- Currently scaling fixes the corner or side of the full bounding box opposite to the anchor you use. Maybe it should fix the corresponding corner of the reduced bounding box (the non-visible one used for snapping). 
- Maybe the rotation center should be the center of the reduced bounding box.
- It would be nice to have the option to keep lines width (e.g. while holding the SHIFT key) for scaled selections.

**Comment by rolandlo** [comment]
The last commit closes issue #2021.

**Comment by rolandlo** [comment]
Now snapping should work for all selections, not just those consisting of strokes.

**Comment by rolandlo** [comment]
Although there may still be room for improvement, this PR seems reasonably complete. So please check it out. 
@frohro @Fatih20 you may want to check, if the snapping behaviour suits your needs.

**Comment by Fatih20** [comment]
> Although there may still be room for improvement, this PR seems reasonably complete. So please check it out.
> @frohro @Fatih20 you may want to check, if the snapping behaviour suits your needs.
I'll download the fork and compile it to test it later, thanks for the heads up!

**Comment by frohro** [comment]
Thanks for your work on this Roland!  Here are my comments:

- I had to do a complete cmake, make, make install after switching to the extend-snapping branch.  i had linker issues if I did not redo the cmake after mistakenly building the master branch.
- The cursor render for resizing is not at the correct angle for the box you select for scaling.  See the video.  It seems to happen after rotating, but I'm not exactly sure yet.  
output

- I agree that it would be nice if scale changes kept the lines the same width, or at least there was an option  for that.  For me it would be the preferred behavior.  As it is, I don't use resizing at all on objects with lines, because of this.
- I agree that resizing snapping would be nice too.
Again, this is a great improvement!  Thanks for all your work Roland.  It will make my online classes better.

Rob

**Comment by rolandlo** [comment]
Thanks @frohro for your feedback and for pointing out a bug with the angle for the cursor. I think I will find an easy fix for that. It should come down to resetting the angle to 0 after a selection has been finalized.

As for the line width I see three ways. 
1) make it configurable in the settings or 
2) have a key modifier (SHIFT or CTRL) or 
3) have both a setting for the usually desired behavior and have a key modifier to change it occasionally.

The downside of 2. is that it is not practical if you work mostly without keyboard. So I lean to going for 1 or 3. 

For resizing snapping: Do you have a clear idea what the behavior should be? I have a clear idea about it when the selection is not rotated, but no idea, if it is rotated. Maybe I could implement resizing snapping only for not rotated selections.

**Comment by rolandlo** [comment]
@frohro: The cursor angle bug is now fixed. If you have time, please check it out again.

**Comment by frohro** [comment]
The cursor bug is fixed for me.

Because xournalpp is in pretty constant flux, and the documentation is really lacking, I don't like Ctrl or Shift key features.  They are really hard to discover by users that aren't reading the code, which is most everyone.  So I vote for option one or three, and it might be nice to have a note in the settings that says Shift or Ctrl swaps the behavior while it is pushed if option three is pursued.

For resize snapping, you could snap the corner or even side closest to what is grabbed to the grid.  That way a user could resize first one side to fit the grid, and then the other, if they want two sides to fit the grid, at the expense of changing the size.  I think the size is the thing that should have the least precedence, since in xournalpp it is all relative anyway.  I really like how you made the snap to grid work for the side or corner closest to what you grabbed.  This would fit in nicely with your already constructed paradigm.

Thanks for all your work on this.  For me at least it is a tremendously useful feature. 

Rob

**Comment by rolandlo** [comment]
Line with scaling can now be enabled or disabled. See in the drawing area settings.

**Comment by frohro** [comment]
Thanks Roland,
I find the setting in the Drawing Area setting, but it behaves the same way with it checked or not.
Rob

**Comment by rolandlo** [comment]
You will only see the difference after finalizing the selection (by clicking outside of it). Before you only see a preview (elements are not updated, only the cairo-surface is transformed). Take a look:
lineWidthDemo

**Comment by frohro** [comment]
Ah!  I like it.  It makes the objects look so much better!  Thanks for the explanation.  We need to think about how to document this feature, as a lot of guys will be like me, and not unselect it until we like it.  :-)

Thanks!!
Rob

**Comment by rolandlo** [comment]
resizing and rotation is now based on the reduced bounding box. This makes it much easier to get snapping for resizing selections as well.
The bug #2034 is still present in the PR and must be fixed there as well.

**Comment by rolandlo** [comment]
Snapping for resizing selections is now implemented. If the selection is not horizontal the snapping acts like the rotation were not there. 
There are some problems with undo actions, in particular bug #2034 
Moreover while resizing the cursor sticks to the reduced bounding box and can be far from the resizing anchors. This may need correction.

Nevertheless snapping should work and may be useful, please check it out, Rob @frohro

**Comment by frohro** [comment]
This looks pretty nice Roland.  I played around with it for a while, and I can get it to do what I want!  It is not quite how I envisioned it, but it works just as well, as I can still make end points snap where I want them to, and make things as large as I like, etc.  I played with it for about half an hour and could find no problems.

**Comment by rolandlo** [comment]
Now undo actions work as expected.

**Comment by rolandlo** [comment]
It is time for some review from one of the active developers @LittleHuba @Febbe @Technius

**Comment by frohro** [comment]
Hi Roland,
There is one more thing I found that needs to snap to grid.  The shape recognizer does not step the recognized shapes to the grid.
Thanks for all your great work on this.  You are making xournalpp much better for my purposes at least.
Rob

**Comment by rolandlo** [comment]
Hi Rob
Great to hear that my efforts to improve snapping help for your purposes. I am aware that the shape recognizer doesn't support snapping yet. It seemed to me (at first sight at least) to be a bit more complicated to do that, so I excluded it from this PR. I will think about it in future, though.
If the shape recognizer should support snapping, then there are two ways to approach it. Either snap the (points defining the) recognized shape to grid. Or try to recognize a shape which is aligned to grid to start with. The second approach is most likely more complicated. In my experience the shape recognizer only recognizes circles, straight lines and rectangles, although there is code in xournalpp supporting more complex shapes. 
Would it suffice for you if snapping for the shape recognizer would only snap recognized straight lines, nearly horizontal or vertical rectangles and circles? 
- in case of straight lines it would simply snap the end points to the grid.
- in case of nearly horizontal or vertical rectangles it would rotate them to be exactly horizontal/vertical and then snap the four corners to the grid
- in case of a circle it would snap the two points on the horizontal diameter to the grid.
Or what would you need it for exactly?

**Comment by frohro** [comment]
Hi Roland,
For me the purpose of shape recognition is just so I don't have to select the right tool to make the shape neat.  
I, likewise, find the only shapes that actually work are the circle, rectangle, and line.  Sometimes I wish the arrow would work too, but I've never been able to make it work.  For me the recognition parameters are not loose enough to get good recognition about half the time, even trying hard to make it neat.  It would be an improvement if there were a sloppiness adjustment in the settings that would make it recognize sloppier shapes.
You described what I want for recognized shapes exactly.  This would make drawing block diagrams and schematics easier.  
Thanks again for all your work on xournalpp!
Rob

**Comment by rolandlo** [comment]
There was a problem with resizing horizontal and vertikal strokes, which is fixed with the latest commit. Moreover I made the cursor stick to the visible bounding box, which improves the ux.

**Comment by rolandlo** [comment]
Hi Rob
Please check out the snapping for shape recognizer. I found a simple concept for it: just snap the reduced bounding box to grid. This works in all cases (line, rectangle, circle) equally well.

**Comment by frohro** [comment]
Thanks Roland!  I played with it for half an hour, trying to practice my box writing, etc.  It is very much better because you can have the items where you want them, but the recognition is still too spotty to make it actually useful.  Circles are easier to draw than boxes, but now sometimes my boxes get recognized as circles.  My issue, now is, even with practice, I make so many figures that are not recognized or not recognized correctly that the time I save by not having to select the tools is outweighed by the undo and redo operations I have to do to correct mistakes.  This addition is a big improvement, though its benefits won't be fully useful until the recognition gets better.  :-)  At least it makes the snap to grid complete, which I feel is a major benefit.

**Comment by rolandlo** [comment]
Thanks for the feedback, Rob!
The shape recognizer code is pretty old. It was written by Denis Auroux, the main author of xournal, and hardly modified since then. The algorithm was documented at [the xournal repository](https://sourceforge.net/p/xournal/support-requests/6/) Some of the recognition features (like triangle recognition) have been disabled in xournalpp. I think, some work should be invested into improving this algorithm. I am not sure that fiddling around with the "sloppiness" constants (seen in ShapeRecognizerConfig.h) will bring much.
I did a quick test, drawing circles, line segments and rectangles by hand, 10 of each kind. From these 10/10 were recognized as circles, 9/10 as line segments, but only 4/10 as rectangles. So in my opinion one should try to improve on the rectangle recognition part. 
I have other plans though for the near future, though.
By the way I never had a box (rectangle) recognized as a circle.

File: src/model/Stroke.cpp
Comment (Febbe) [inline src/model/Stroke.cpp:186]: Just some notes, for me the functions needs to be evaluated for a refactor:
 - doesn't fx*fy == 1 imply, that doScale is false?
 - it look some kind of odd, that there is a rotation used to scale, this should be irrelevant, when previous translations are applied, or the element childs are positioned relative to to the element, isn't it?
 - the thickness `fz` should be calculated from the angle it has (Law of cosines), not just the pythagorean theorem. Currently the scaling is only correct for perfectly diagonal lines.
 - also this function looks more like a general translation function, maybe renaming it / creating a new would be more clear.

File: src/model/Text.cpp
Comment (Febbe) [inline src/model/Text.cpp:129]: add a newline

File: src/model/Text.cpp
Comment (Febbe) [inline src/model/Text.cpp:78]: call calculateSize before this line or remove it, since the width and height is not calculated yet.

File: src/control/settings/Settings.cpp
Comment (Febbe) [inline src/control/settings/Settings.cpp:161]: You changed the default behaviour of xournal++, this will be very unexpected, resizing written text is not possible without changing it back.

**Comment by rolandlo** [comment]
Hi @febbe, thanks a lot for your helpful comments and explanations.

File: src/control/settings/Settings.cpp
Comment (rolandlo) [reply src/control/settings/Settings.cpp:161]: Ok, I will set this to true as default.

File: src/model/Stroke.cpp
Comment (rolandlo) [reply src/model/Stroke.cpp:186]: To scale the line width and pressure of points with the geometric mean of the scaling factors was not my idea, but makes some sense to me.

- In the special case where fx*fy==1 then the line width and pressure of points is left unmodified. The boolean doScale is about whether line width and pressure should be potentially modified after a selection has been resized.
- The rotation angle for this generalized scale function is only used for the Undo Handler (see ScaleUndoAction.cpp:46). The Undo Handler does not remember how the selection was rotated, when the scaling was done, it just remembers what it should undo/redo with the elements of the selection. Note that this function does scale in a rotated coordinate system. So instead of scaling coordinates by fx in direction of the usual x-axis and by fy in direction of the usual y-axis, it does scaling with respect to the rotated x- and y-axis.
- I think the thickness fz should be the same for the whole stroke, not depend on the angle of the single line segments. Otherwise, if you want to do it with "correct scaling", you get ugly strokes. This is already visible if you scale a rectangle (with round line caps) only in x-direction (or y-direction). See the following example:
grafik

- If you prefer, I can remove the rotation and change the undo / redo action. However the rotation function also uses a rotation center, so it is also not just standard rotation around the origin.

File: src/model/Image.cpp
Comment (rolandlo) [reply src/model/Image.cpp:86]: I know, but rotation may also be used by the Image class eventually. The boolean doScale will never be used, though. It is a virtual function, so all derived classes (from the base class Element) must use the same arguments, don't they? How should I go about it, if the argument doScale is only used by the derived class Stroke?

File: src/model/Element.cpp
Comment (rolandlo) [reply src/model/Element.cpp:54]: I don't know yet what to do about that. We have the same issue with the unsnapped values as well. Those getters are called quite often and on various places. The snapped values are rarely called, in contrast.

**Comment by rolandlo** [comment]
For some requested changes (about getters and about the scaling function) I have formulated questions (see above) and will wait for some feedback. The other changes are implemented.

File: src/model/Element.cpp
Comment (rolandlo) [reply src/model/Element.cpp]: I tried to make all getters related to the bounding boxes const by making all cache variables mutable and all functions called by the getters const, but still failed. The compiler complained about the static function TextView::calcSize()

**Comment by rolandlo** [comment]
@Febbe: Now I have made all requested changes except those you suggested for the scaling function (see my questions above). Tell me, if you still want me to change that after seeing my comments.

**Comment by frohro** [comment]
Roland,
I was using your work this morning making figures for python notebooks, and snapping moved objects to the grid is just fantastic.  Thanks for all the time you put into this.  It really helps make xournalpp great for my applications.  I can now copy and paste repeated parts of the figure easily, and they end up where I want them.  It was so nice!
Thanks!
Rob

**Comment by frohro** [comment]
I've been waiting for this to be accepted since May.  In the meantime, I have been using @rolandlo's fork that has these changes.  I have been using it a lot.  Just today, I used the ability to select a picture and move it over a certain number of squares in my Signals & Systems class.  it was great!  I hope it doesn't molder until merging it is a problem.  Operationally, it is great!

**Comment by Febbe** [comment]
Thanks frohro for the reminder. @rolandlo, you have a better overview, what's missing in your repository before we merge it. ( You should squash some commits together, but what else is open?)

**Comment by rolandlo** [comment]
The only real thing that comes to mind: The snapping for recognized shapes may need to be made optional. I often have the shape recognizer (and snapping to grid) turned on even if I just write some text, so that lines are straight if I underline a title (or something else). Then I don't want this line to be snapped to grid.

Maybe rotation snapping could be made for 15 degree steps (currently it is 45 degrees). That would allow for easy creation of equilateral triangles and things like that. I am not sure whether that should be made configurable in the settings or not. It can wait for a follow-up PR as well.

Otherwise I don't see why it shouldn't be merged. Seems @frohro  has done quite a bit of testing and it works as expected.

So @Febbe, if you agree, I will make snapping for recognized settings optional by adding a checkbox in the drawing area panel in the settings, and then squash commits together.

**Comment by Febbe** [comment]
Maybe a switch between offset snapping as extension, instead of snapping to the grid. That would allow to start shapes exact at the pen position, and the snapping is relative to the first point. If the position is bad someone can still snap it to the grid by moving the object.

**Comment by rolandlo** [comment]
That offset switch idea sounds like an interesting suggestion, but most likely not trivial to implement and requiring decent testing. I would prefer to leave that for a follow-up PR and finish this PR up rather sooner than later.

**Comment by rolandlo** [comment]
I have made the changes I wanted (configurable snapping for shape-recognizer, 15 degree rotation snapping). Some (minor) merge conflicts have to be solved. Seems I have messed up with git trying to squash the commits and solve the merge conflicts.

**Review by Febbe — APPROVED** [review]

File: src/model/Snapping.cpp
Comment (rolandlo) [reply src/model/Snapping.cpp:18]: I know, but I think these methods will diverge in future if we introduce snapping to other grids. Should I still make 1 function out of it?

**Comment by rolandlo** [comment]
Seems I have managed to squash and rebase on current master, resolving all merge conflicts.

**Comment by Febbe** [comment]
Ok can you squash the last commit. And then you can merge it.

**Comment by rolandlo** [comment]
Thanks a lot Febbe for taking the time to look into it. Wonderful that this is finished now.

**Pull request merged**
